weave/utils/iterators.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+from collections.abc import Generator, Iterator, Sequence
+from threading import Lock
+from typing import Any, TypeVar, overload
+
+T = TypeVar("T")
+
+
+class ThreadSafeInMemorySequence(Sequence[T]):
```

Comment:
As I read through the code, I think more and more of this as "LazyList"

Reply:
yeah.... this is just a lazylist

Reply:
or maybe `ThreadSafeLazyList`